File: packages/queries/github.ts
import type { PullRequestResult } from '@openflow/generated';
import { createLogger } from '@openflow/utils';
import { invoke } from './utils.js';

/**
 * Logger for GitHub query operations.
 * Logs at DEBUG level for query calls, INFO for successes, ERROR for failures.
 */
const logger = createLogger('queries:github');

/**
 * Input for creating a pull request.
 */
export interface CreatePullRequestInput {
  /** The task ID to create a PR for */
  taskId: string;
  /** PR title (defaults to task title if not provided) */
  title?: string;
  /** PR body/description */
  body?: string;
  /** Base branch to merge into (defaults to project's base_branch or "main") */
  base?: string;
  /** Whether to create as a draft PR */
  draft?: boolean;
}

/**
 * GitHub query wrappers for Tauri IPC.
 * Thin wrappers around invoke() calls for GitHub operations.
 *
 * All functions include:
 * - Try/catch error handling with re-throw for React Query
 * - Logging at appropriate levels (DEBUG on call, INFO on success, ERROR on failure)
 *
 * @example
 * ```ts
 * // Check if GitHub CLI is available
 * const isInstalled = await githubQueries.checkGhCliInstalled();
 *
 * // Check authentication status
 * await githubQueries.checkGhAuthStatus();
 *
 * // Create a pull request
 * const result = await githubQueries.createPullRequest({
 *   taskId: 'task-123',
 *   title: 'Add new feature',
 *   body: 'This PR adds...',
 *   draft: false,
 * });
 * console.log(`PR created: ${result.url}`);
 * ```
 */
export const githubQueries = {
  /**
   * Create a pull request for a task using the GitHub CLI.
   *
   * This will:
   * 1. Verify the GitHub CLI is installed and authenticated
   * 2. Get the task's worktree path
   * 3. Push the branch to the remote
   * 4. Create the PR using `gh pr create`
   *
   * @param input - Pull request creation options
   * @returns The result containing the PR URL, number, and branch name
   * @throws Error if the query fails (re-thrown for React Query)
   */
  createPullRequest: async (input: CreatePullRequestInput): Promise<PullRequestResult> => {
    logger.debug('Creating pull request', {
      taskId: input.taskId,
      hasTitle: !!input.title,
      hasBody: !!input.body,
      base: input.base,
      draft: input.draft,
    });

    try {
      const result = await invoke<PullRequestResult>('create_pull_request', { input });

      logger.info('Pull request created successfully', {
        taskId: input.taskId,
        prNumber: result.number,
        prUrl: result.url,
        branchName: result.branch,
        isDraft: input.draft,
      });

      return result;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to create pull request', {
        taskId: input.taskId,
        title: input.title,
        base: input.base,
        draft: input.draft,
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Check if the GitHub CLI is installed and available.
   *
   * @returns `true` if the `gh` CLI is installed and executable
   * @throws Error if the query fails (re-thrown for React Query)
   */
  checkGhCliInstalled: async (): Promise<boolean> => {
    logger.debug('Checking if GitHub CLI is installed');

    try {
      const isInstalled = await invoke<boolean>('check_gh_cli_installed', {});

      logger.info('GitHub CLI installation check completed', {
        isInstalled,
      });

      return isInstalled;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('Failed to check GitHub CLI installation', {
        error: errorMessage,
      });
      throw error;
    }
  },

  /**
   * Check if the user is authenticated with GitHub CLI.
   *
   * @throws Error if not authenticated or if the query fails (re-thrown for React Query)
   */
  checkGhAuthStatus: async (): Promise<void> => {
    logger.debug('Checking GitHub CLI authentication status');

    try {
      await invoke<void>('check_gh_auth_status', {});

      logger.info('GitHub CLI authentication check passed');
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.error('GitHub CLI authentication check failed', {
        error: errorMessage,
      });
      throw error;
    }
  },
};

File: src-tauri/src/commands/github.rs
//! Tauri commands for GitHub operations.
//!
//! These commands provide the IPC interface for GitHub-related operations,
//! primarily pull request creation using the GitHub CLI.

use tauri::State;

use crate::commands::AppState;
use crate::services::GitHubService;
use crate::types::PullRequestResult;

/// Request structure for creating a pull request.
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePullRequestInput {
    /// The task ID to create a PR for.
    pub task_id: String,
    /// PR title (defaults to task title if not provided).
    pub title: Option<String>,
    /// PR body/description.
    pub body: Option<String>,
    /// Base branch to merge into (defaults to project's base_branch or "main").
    pub base: Option<String>,
    /// Whether to create as a draft PR.
    #[serde(default)]
    pub draft: bool,
}

/// Create a pull request for a task using the GitHub CLI.
///
/// This command:
/// 1. Verifies the GitHub CLI is installed and authenticated
/// 2. Gets the task's worktree path
/// 3. Pushes the branch to the remote
/// 4. Creates the PR using `gh pr create`
///
/// # Arguments
/// * `task_id` - The task to create a PR for
/// * `title` - PR title (defaults to task title)
/// * `body` - PR body/description
/// * `base` - Base branch to merge into (defaults to project's base_branch or "main")
/// * `draft` - Whether to create as a draft PR
///
/// # Returns
/// The result containing the PR URL, number, and branch name.
///
/// # Errors
/// Returns an error if:
/// - GitHub CLI is not installed
/// - User is not authenticated with GitHub
/// - Task has no active worktree
/// - Push to remote fails
/// - PR creation fails
#[tauri::command]
pub async fn create_pull_request(
    state: State<'_, AppState>,
    input: CreatePullRequestInput,
) -> Result<PullRequestResult, String> {
    let pool = state.db.lock().await;

    GitHubService::create_pull_request(
        &pool,
        &input.task_id,
        input.title,
        input.body,
        input.base,
        input.draft,
    )
    .await
    .map_err(|e| e.to_string())
}

/// Check if the GitHub CLI is installed and available.
///
/// # Returns
/// `true` if the `gh` CLI is installed and executable.
#[tauri::command]
pub fn check_gh_cli_installed() -> Result<bool, String> {
    GitHubService::check_gh_cli_installed().map_err(|e| e.to_string())
}

/// Check if the user is authenticated with GitHub CLI.
///
/// # Returns
/// Empty result on success, error message if not authenticated.
#[tauri::command]
pub fn check_gh_auth_status() -> Result<(), String> {
    GitHubService::check_gh_auth_status().map_err(|e| e.to_string())
}
